swap/simulations_test.go

```diff
@@ -325,18 +327,20 @@ func TestPingPongChequeSimulation(t *testing.T) {
 			if err := p2Peer.Send(context.Background(), &testMsgBigPrice{}); err != nil {
 				t.Fatal(err)
 			}
-			if err := waitForChequeProcessed(t, ts2, counter, lastCount); err != nil {
+			if err := waitForChequeProcessed(t, params.backend, counter, lastCount, ts1.swap.peers[p2], expectedPayout1); err != nil {
 				t.Fatal(err)
 			}
 			lastCount += 1
+			expectedPayout1 += DefaultPaymentThreshold + 1
 		} else {
 			if err := p1Peer.Send(context.Background(), &testMsgBigPrice{}); err != nil {
```

Comment:
In one test run (out of 2000) I got a nil pointer in this line:
```
   --- FAIL: TestPingPongChequeSimulation (0.08s)
   panic: runtime error: invalid memory address or nil pointer dereference [recovered]
   panic: runtime error: invalid memory address or nil pointer dereference
   [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xc52f37]
       
   goroutine 36 [running]:
   testing.tRunner.func1(0xc0004b5300)
   	/usr/local/go/src/testing/testing.go:874 +0x3a3
   panic(0xd84600, 0x16a17b0)
   	/usr/local/go/src/runtime/panic.go:679 +0x1b2
   github.com/ethersphere/swarm/swap.TestPingPongChequeSimulation(0xc0004b5300)
   	/home/user/swarm/swap/simulations_test.go:336 +0x927
   testing.tRunner(0xc0004b5300, 0xf3c3f0)
   	/usr/local/go/src/testing/testing.go:909 +0xc9
   created by testing.(*T).Run
   	/usr/local/go/src/testing/testing.go:960 +0x350
   FAIL	github.com/ethersphere/swarm/swap	0.161s
   FAIL
```

Reply:
Hmmm that's a hell of a difficult one. but I also wonder how many tests in the entire code base would stand the check of 1000 runs. It never happened before in many high count runs...

Reply:
From what I see the only thing that could be `nil` here is `p1Peer` (which should not be nil because we wait for swap connections to be established). Maybe the protocol got disconnected inbetween the waiting loop and `p1Peer := ts2.peers[p1] `. Either way if it doesn't happen again we can probably ignore this. Also very unlikely that this is related to the changes from this PR.

Reply:
One thing that may be related to this is that there are two maps of peers in this test. One is `testService.peers` and another `testService.swap.peers`. The panic is on the peer from the first one but the for loop is checking for the second one if it contains at least one peer. I think that it would be more correct to wait for the first peers map:

```go
		ts1.lock.Lock()
		ts1Len = len(ts1.peers)
		ts1.lock.Unlock()

		ts2.lock.Lock()
		ts2Len = len(ts2.peers)
		ts2.lock.Unlock()
```

This may fix this issue as the nil peer can happen if the map does not contain it. But I did not manage to confirm that.

Reply:
The reason why we wait for the 2nd map is that we only can know that we have all connections established when the 2nd map has the expected amount of connections, otherwise when sending cheques we may not know if the peer is there already.

So maybe the real thing is to wait for both. I am thinking that we could to that under the same lock,,,I will do some trial but if it doesn't work well I will leave it as-is.